### ISSUE TITLE
Debug Printf produces no output for shaders created via VK_EXT_shader_object (silent failure, no error)

### ISSUE THREAD

**Issue by mklefrancois** [KhronosGroup/Vulkan-ValidationLayers#12597]
## Environment:

GPU: NVIDIA RTX 3500 Ada Generation Laptop GPU, driver 610.188.0
Vulkan API version: 1.4.350
Khronos Validation Layer: v1.4.350
OS: Windows 11 (10.0.26200)

## Description:

VK_VALIDATION_FEATURE_ENABLE_DEBUG_PRINTF_EXT correctly instruments shaders bound via a classic VkPipeline, producing printf output as expected. The identical SPIR-V, when instead built into VkShaderEXT objects via vkCreateShadersEXT and bound with vkCmdBindShadersEXT, produces no printf output at all — not even the raw message before any application-side filtering (tested with khronos_validation.printf_to_stdout=true, which bypasses app-level message handling entirely). No validation errors or warnings are reported either — it fails completely silently.

This may be related to [#7178](https://github.com/KhronosGroup/Vulkan-ValidationLayers/issues/7178), which shows GPU-AV injecting a descriptor set (index #31) for VK_EXT_shader_object instrumentation and hitting a false-positive compatibility error in the process. Our case differs: no error at all, just zero output — possibly this is what remains after whatever fixed [#7178](https://github.com/KhronosGroup/Vulkan-ValidationLayers/issues/7178)'s false positive, if the fix narrowed or disabled the injection for shader objects rather than completing it. Filing as a separate report since the symptoms don't match closely enough to be confident it's the same root cause.

## Repro steps:

1. Minimal vertex+fragment shader pair, fragment shader has one conditional printf(...) (Slang) / debugPrintfEXT(...) (GLSL) call, guarded by a condition guaranteed true for at least one covered fragment.
2. Enable VK_VALIDATION_FEATURE_ENABLE_DEBUG_PRINTF_EXT, set printf_to_stdout=true.
3. A — classic pipeline: build via vkCreateGraphicsPipelines from the SPIR-V, bind with vkCmdBindPipeline, draw → printf output appears every frame the condition is met.
4. B — shader objects: same SPIR-V, vkCreateShadersEXT (VK_SHADER_CREATE_LINK_STAGE_BIT_EXT, VK_SHADER_CODE_TYPE_SPIRV_EXT), bind via vkCmdBindShadersEXT with dynamic state from VK_EXT_extended_dynamic_state3, same draw → zero output, headless or windowed, reproduced across multiple independent test runs.
5. Rendering itself is correct in both paths (verified visually) — only the debug-printf instrumentation differs.

**Comment by spencer-lunarg** [comment]
>  No validation errors or warnings are reported either — it fails completely silently.

> no error at all, just zero output 

... these are conflicting, what is actually happening for you?

#7178 is old and I looked, we now have 17 different tests now in `tests/unit/debug_printf.cpp` that use shader object, and they are doing all sort of different things

Without some way to reproduce, or at least provide the API dump of the `vkCreateShadersEXT`/`vkCmdBindShadersEXT` (in case there is something you are doing special) I have no idea why this is not working for you

**Comment by mklefrancois** [comment]
Hi @spencer-lunarg  — following up on this, and sorry for the noise here.

You were right to push back — I went back and built a completely minimal, standalone repro (no framework, single file) to isolate this properly, and the "shader objects break Debug Printf" framing was wrong from the start. So was my follow-up claim that vkCmdPushDataEXT itself was the cause.

The actual root cause was on my end: the shader objects were never created with VK_SHADER_CREATE_DESCRIPTOR_HEAP_BIT_EXT. Once I add that flag to the VkShaderCreateInfoEXT for shaders that receive data via vkCmdPushDataEXT, Debug Printf works correctly, every time. I confirmed this with a single-variable A/B: identical shader objects, identical bind, identical draw — toggling only that one flag is the difference between zero printf output and printf firing reliably.

This is entirely my mistake, and I apologize for taking up your time with an incorrect report.

One thing that might still be worth a look on your end, if you're open to it: there was no validation error or warning anywhere in the process pointing at the missing flag. The only VUID I found related to it (VUID-VkShaderCreateInfoEXT-flags-11370) only fires in the other direction — checking pushConstantRangeCount == 0 if the flag is set — but nothing catches the case where a shader lacking the flag is later used with vkCmdPushDataEXT. A stateless check there (something like "shader bound at vkCmdPushDataEXT/draw time must have been created with VK_SHADER_CREATE_DESCRIPTOR_HEAP_BIT_EXT") would have caught my mistake immediately instead of silently dropping instrumentation output. Feel free to close this either way — happy to file that as a separate, much smaller suggestion if useful, otherwise no action needed.

Thanks again for catching this early and saving everyone the trouble of chasing a phantom bug.

**Comment by spencer-lunarg** [comment]
`vkCmdPushDataEXT` + `ShaderObject` is something we don't have a test for... let me look into this! (yay API permutation hell)

**Comment by mklefrancois** [comment]
If this can be useful, here is a minimal reproducing example. 
To repro the bad behavior, comment the line:

```cpp
  shaderFlags |= VK_SHADER_CREATE_DESCRIPTOR_HEAP_BIT_EXT;
```


[khronos_vvl_repro.zip](https://github.com/user-attachments/files/29611796/khronos_vvl_repro.zip)

**Comment by spencer-lunarg** [comment]
@mklefrancois just got back to this... I see we have ` VUID-vkCmdDraw-None-08601` and `VUID-vkCmdDraw-None-11376` but I think you just found a simple, but clearly easy to mess up, VU gap coverage

will get coverage and add a new VU to the spec

**Comment by spencer-lunarg** [comment]
ok, another update, I realize the issue is you called `vkCmdPushDataEXT` but the shader never uses any descriptors nor push constant data

- if your shader would have used `layout(push_constant)` inside, you would have hit either `08601` or `11376`
- if your shader would have used a descriptor, you would have hit other errors
- If tried with Pipelines instead of ShaderObjects, you would have hit 07826

... really what you hit was DebugPrintf needs to know how to handle this shader, and we saw `vkCmdPushDataEXT` and thought "ok... this is a descriptor heap" but really the data is never used and we could have gone "ok... the flag was never set"

For this, I think at best we could give a warning (as it is valid and will 100% work), but really could just be something debugprintf just handles correctly

**Comment by spencer-lunarg** [comment]
repro case

```
TEST_F(NegativeDebugPrintf, DescriptorHeapShadeerObjectUselessPushData) {
    TEST_DESCRIPTION("https://github.com/KhronosGroup/Vulkan-ValidationLayers/issues/12597");
    SetTargetApiVersion(VK_API_VERSION_1_3);
    AddRequiredExtensions(VK_EXT_DESCRIPTOR_HEAP_EXTENSION_NAME);
    AddRequiredExtensions(VK_EXT_SHADER_OBJECT_EXTENSION_NAME);
    AddRequiredFeature(vkt::Feature::shaderObject);
    AddRequiredFeature(vkt::Feature::dynamicRendering);
    AddRequiredFeature(vkt::Feature::descriptorHeap);
    AddRequiredFeature(vkt::Feature::bufferDeviceAddress);
    AddRequiredFeature(vkt::Feature::vertexPipelineStoresAndAtomics);
    RETURN_IF_SKIP(InitDebugPrintfFramework());
    RETURN_IF_SKIP(InitState());
    InitDynamicRenderTarget();

    const char* vs_source = R"glsl(
        #version 450
        #extension GL_EXT_debug_printf : enable
        void main() {
            if (gl_VertexIndex == 0) {
                debugPrintfEXT("x == %u\n", gl_VertexIndex);
            }
            gl_Position = vec4(1.0f);
        }
    )glsl";

    const auto vert_spv = GLSLToSPV(VK_SHADER_STAGE_VERTEX_BIT, vs_source);
    const auto frag_spv = GLSLToSPV(VK_SHADER_STAGE_FRAGMENT_BIT, kFragmentMinimalGlsl);

    // not using ShaderCreateInfoHeap()
    VkShaderCreateInfoEXT vert_ci = ShaderCreateInfoLink(vert_spv, VK_SHADER_STAGE_VERTEX_BIT, VK_SHADER_STAGE_FRAGMENT_BIT);
    VkShaderCreateInfoEXT frag_ci = ShaderCreateInfoLink(frag_spv, VK_SHADER_STAGE_FRAGMENT_BIT);
    vkt::Shader vert_shader(*m_device, vert_ci);
    vkt::Shader frag_shader(*m_device, frag_ci);

    m_command_buffer.Begin();
    m_command_buffer.BeginRenderingColor(GetDynamicRenderTarget(), GetRenderTargetArea());
    SetDefaultDynamicStatesExclude();
    const VkShaderStageFlagBits stages[] = {VK_SHADER_STAGE_VERTEX_BIT, VK_SHADER_STAGE_FRAGMENT_BIT};
    vk::CmdBindShadersEXT(m_command_buffer, 1u, &stages[0], &vert_shader.handle());
    vk::CmdBindShadersEXT(m_command_buffer, 1u, &stages[1], &frag_shader.handle());
    uint32_t dummy_data = 0;
    m_command_buffer.PushData(0, 4, &dummy_data);
    vk::CmdDraw(m_command_buffer, 3, 1, 0, 0);
    m_command_buffer.EndRendering();
    m_command_buffer.End();

    m_errorMonitor->SetDesiredInfo("x == 0");
    m_default_queue->SubmitAndWait(m_command_buffer);
    m_errorMonitor->VerifyFound();
}
```

**Comment by spencer-lunarg** [comment]
https://gitlab.khronos.org/vulkan/vulkan/-/issues/4896

**Comment by mklefrancois** [comment]
Thanks for filing the spec issue. I'm glad to see this being worked on. I'm happy to test against a patched layer build whenever you have coverage in place.
